File: solution/f#/Nir-Kata-Tests/Parse.Dont.Validate/NIRGenerator.fs
module NirGenerator

open FsCheck
open Nir_Kata.Parse.Dont.Validate.NIRDomain.City
open Nir_Kata.Parse.Dont.Validate.NIRDomain.Department
open Nir_Kata.Parse.Dont.Validate.NIRDomain.Month
open Nir_Kata.Parse.Dont.Validate.NIRDomain.NIRDomain
open Nir_Kata.Parse.Dont.Validate.NIRDomain.SerialNumber
open Nir_Kata.Parse.Dont.Validate.NIRDomain.Sex
open Nir_Kata.Parse.Dont.Validate.NIRDomain.Year

type NirGenerator =
    static member Nir() : Arbitrary<NIR> =
        gen {
            let! sex = Arb.generate<Sex>
            let! year = Gen.choose (0, 99)
            let! month = Gen.choose (1, 12)

            let! department =
                Gen.frequency [ (9, Gen.choose (1, 95))
                                (1, Gen.constant 99) ]

            let! city = Gen.choose (1, 999)
            let! serialNumber = Gen.choose (1, 999)

            return
                { sex = sex
                  year = Year year
                  month = Month month
                  department = Department department
                  city = City city
                  serialNumber = SerialNumber serialNumber }
        }
        |> Arb.fromGen

File: solution/c#/Nir-Kata-Tests/Parse.Dont.Validate/NIRGenerator.cs
using System.Diagnostics.CodeAnalysis;
using FsCheck;
using Nir_Kata.Parse.Dont.Validate;
using static System.Tuple;

namespace Nir_Kata_Tests.Parse.Dont.Validate
{
    public static class NIRGenerator
    {
        [SuppressMessage("ReSharper", "UnusedMember.Local", Justification = "Used by FSCheck")]
        public static Arbitrary<NIR> Nir() =>
            (from sex in Arb.Generate<Sex>()
                from year in Gen.Choose(0, 99)
                from month in Arb.Generate<Month>()
                from department in Gen.Frequency(new[] {Create(9, Gen.Choose(1, 95)), Create(1, Gen.Constant(99))})
                from city in Gen.Choose(1, 999)
                from serialNumber in Gen.Choose(1, 999)
                select new NIR(sex, (Year) year, month, (Department) department, (City) city,
                    (SerialNumber) serialNumber)
            ).ToArbitrary();
    }
}